Assets/MixedRealityToolkit-SDK/Features/Diagnostics/MixedRealityDiagnosticsManager.cs

```diff
@@ -0,0 +1,243 @@
+                showFps = false;
+                showMemory = false;
+
+                RaiseDiagnosticsChanged();
+            }
+        }
+
+        public GameObject GetDiagnosticVisualization()
+        {
+            if (diagnosticVisualization != null)
+            {
+                return diagnosticVisualization;
+            }
+
+            if (!Visible)
+            {
+                // Don't create a gameobject if it's not needed
+                return null;
+            }
+
+            diagnosticVisualization = GameObject.CreatePrimitive(PrimitiveType.Quad);
+            diagnosticVisualization.name = "Diagnostics";
+            diagnosticVisualization.layer = 2; // magic? 
+
+            // Todo: position and size
+            // Todo: add text elements
+            diagnosticVisualization.AddComponent<DiagnosticBehavior>();
+
+            var component = diagnosticVisualization.GetComponent<DiagnosticBehavior>();
+            component.ShowCpu = ShowCpu;
+            component.ShowFps = ShowFps;
+            component.ShowMemory = ShowMemory;
+
+            return diagnosticVisualization;
+        }
+
+        private void RaiseDiagnosticsChanged()
+        {
+            eventData.Initialize(this,
+                visible: Visible,
+                showCpu: ShowCpu,
+                showFps: ShowFps,
+                showMemory: ShowMemory
+                );
+
+            HandleEvent(eventData, OnDiagnosticsChanged);
+        }
+
+        /// <summary>
+        /// Event sent whenever the boundary visualization changes.
+        /// </summary>
+        private static readonly ExecuteEvents.EventFunction<IMixedRealityDiagnosticsHandler> OnDiagnosticsChanged =
+            delegate (IMixedRealityDiagnosticsHandler handler, BaseEventData eventData)
+            {
+                DiagnosticsEventData diagnosticsEventsData = ExecuteEvents.ValidateEventData<DiagnosticsEventData>(eventData);
+                handler.OnDiagnosticSettingsChanged(diagnosticsEventsData);
+            };
+
+        #endregion
+
+        #region IMixedRealityDiagnosticsManager
+        private bool visible;
+
+        /// <inheritdoc />
+        public bool Visible
+        {
+            get
+            {
+                return visible;
+            }
+
+            set
+            {
+                if (value != visible)
+                {
+                    visible = value;
+                    GetDiagnosticVisualization()?.SetActive(value);
+
+                    RaiseDiagnosticsChanged();
+                }
+            }
+        }
+
+        private bool showCpu;
+
+        /// <inheritdoc />
+        public bool ShowCpu
+        {
+            get
+            {
+                return showCpu;
+            }
+
+            set
+            {
+                if (value != showCpu)
+                {
+                    showCpu = value;
+                    var component = GetDiagnosticVisualization()?.GetComponent<DiagnosticBehavior>();
```

Comment:
can we cache some of these `GetComponent` calls? They're pretty expensive, and we can always cache it when we set it up in enable

Reply:
Definitely. I'll make the component a member variable too to save the reference. Just need to be careful with lifetime I think

Reply:
If you use the handler, you won't need to keep a cached value. Just raise the event.